### ISSUE TITLE
macOS build failure for 1.34.0 due to C++14

### ISSUE THREAD

**Issue by ilovezfs** [aria2/aria2#1198]
Log https://gist.github.com/ilovezfs/ff000f741d3c745bb36b67615a105c54

The error is
```
bash-3.2$ make V=1
/Library/Developer/CommandLineTools/usr/bin/make  all-recursive
Making all in po
Making all in lib
make[2]: Nothing to be done for `all'.
Making all in deps
Making all in wslay
/Library/Developer/CommandLineTools/usr/bin/make  all-recursive
Making all in lib
Making all in includes
make[6]: Nothing to be done for `all'.
make[6]: Nothing to be done for `all-am'.
Making all in tests
make[5]: Nothing to be done for `all'.
make[5]: Nothing to be done for `all-am'.
make[3]: Nothing to be done for `all-am'.
Making all in src
Making all in includes
make[3]: Nothing to be done for `all'.
/bin/sh ../libtool  --tag=CXX   --mode=compile /usr/bin/clang++ -DHAVE_CONFIG_H -I. -I..  -I../lib -I../intl -I./includes -I./includes -DLOCALEDIR=\"/usr/local/Cellar/aria2/1.34.0/share/locale\" -DHAVE_CONFIG_H -I/usr/include/libxml2 -I../deps/wslay/lib/includes -I../deps/wslay/lib/includes  -F/usr/local/Frameworks -std=c++11   -pipe -Os -w -pipe -march=native -mmacosx-version-min=10.11 -c -o InternalDHKeyExchange.lo InternalDHKeyExchange.cc
libtool: compile:  /usr/bin/clang++ -DHAVE_CONFIG_H -I. -I.. -I../lib -I../intl -I./includes -I./includes -DLOCALEDIR=\"/usr/local/Cellar/aria2/1.34.0/share/locale\" -DHAVE_CONFIG_H -I/usr/include/libxml2 -I../deps/wslay/lib/includes -I../deps/wslay/lib/includes -F/usr/local/Frameworks -std=c++11 -pipe -Os -w -pipe -march=native -mmacosx-version-min=10.11 -c InternalDHKeyExchange.cc  -fno-common -DPIC -o .libs/InternalDHKeyExchange.o
In file included from InternalDHKeyExchange.cc:36:
In file included from ./InternalDHKeyExchange.h:40:
./bignum.h:34:25: error: use of undeclared identifier 'make_unique'
  inline ulong() : buf_(make_unique<char_t[]>(dim)) {}
                        ^
./bignum.h:34:37: error: unexpected type name 'char_t': expected expression
  inline ulong() : buf_(make_unique<char_t[]>(dim)) {}
                                    ^
./bignum.h:34:44: error: expected expression
  inline ulong() : buf_(make_unique<char_t[]>(dim)) {}
                                           ^
./bignum.h:35:33: error: use of undeclared identifier 'make_unique'
  inline ulong(size_t t) : buf_(make_unique<char_t[]>(dim))
                                ^
./bignum.h:35:45: error: unexpected type name 'char_t': expected expression
  inline ulong(size_t t) : buf_(make_unique<char_t[]>(dim))
                                            ^
./bignum.h:35:52: error: expected expression
  inline ulong(size_t t) : buf_(make_unique<char_t[]>(dim))
                                                   ^
./bignum.h:39:46: error: use of undeclared identifier 'make_unique'
  inline ulong(const ulong<dim>& rhs) : buf_(make_unique<char_t[]>(dim))
                                             ^
./bignum.h:39:58: error: unexpected type name 'char_t': expected expression
  inline ulong(const ulong<dim>& rhs) : buf_(make_unique<char_t[]>(dim))
                                                         ^
./bignum.h:39:65: error: expected expression
  inline ulong(const ulong<dim>& rhs) : buf_(make_unique<char_t[]>(dim))
                                                                ^
./bignum.h:44:14: error: use of undeclared identifier 'make_unique'
      : buf_(make_unique<char_t[]>(dim))
             ^
./bignum.h:44:26: error: unexpected type name 'char_t': expected expression
      : buf_(make_unique<char_t[]>(dim))
                         ^
./bignum.h:44:33: error: expected expression
      : buf_(make_unique<char_t[]>(dim))
                                ^
InternalDHKeyExchange.cc:58:12: error: no matching constructor for initialization of 'n' (aka 'ulong<1024>')
  prime_ = n(pr.c_str(), pr.length());
           ^ ~~~~~~~~~~~~~~~~~~~~~~~
./bignum.h:24:29: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
template <size_t dim> class ulong {
                            ^
./bignum.h:24:29: note: candidate constructor (the implicit default constructor) not viable: requires 0 arguments, but 2 were provided
InternalDHKeyExchange.cc:68:16: error: no matching constructor for initialization of 'n' (aka 'ulong<1024>')
  generator_ = n(gen.c_str(), gen.length());
               ^ ~~~~~~~~~~~~~~~~~~~~~~~~~
./bignum.h:24:29: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
template <size_t dim> class ulong {
                            ^
./bignum.h:24:29: note: candidate constructor (the implicit default constructor) not viable: requires 0 arguments, but 2 were provided
InternalDHKeyExchange.cc:73:17: error: no matching constructor for initialization of 'n' (aka 'ulong<1024>')
  privateKey_ = n(reinterpret_cast<char*>(buf), pbytes);
                ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./bignum.h:24:29: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
template <size_t dim> class ulong {
                            ^
./bignum.h:24:29: note: candidate constructor (the implicit default constructor) not viable: requires 0 arguments, but 2 were provided
InternalDHKeyExchange.cc:117:5: error: no matching constructor for initialization of 'n' (aka 'ulong<1024>')
  n peerKey(reinterpret_cast<const char*>(peerPublicKeyData),
    ^       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./bignum.h:24:29: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
template <size_t dim> class ulong {
                            ^
./bignum.h:24:29: note: candidate constructor (the implicit default constructor) not viable: requires 0 arguments, but 2 were provided
16 errors generated.
make[3]: *** [InternalDHKeyExchange.lo] Error 1
make[2]: *** [all-recursive] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
bash-3.2$
```

The build seems to require C++14 not C++11. I can hack around the failure with
```
diff --git a/configure b/configure
index a3a6feb..194e509 100755
--- a/configure
+++ b/configure
@@ -18392,7 +18392,7 @@ $as_echo "$ax_cv_cxx_compile_cxx11" >&6; }
 
 
     if test x$ac_success = xno; then
-    for switch in -std=c++11 -std=c++0x; do
+    for switch in -std=c++14 -std=c++0x; do
       for lib in "" -stdlib=libc++; do
         cachevar=`$as_echo "ax_cv_cxx_compile_cxx11_$switch$lib" | $as_tr_sh`
         { $as_echo "$as_me:${as_lineno-$LINENO}: checking whether $CXX supports C++11 features with $switch $lib" >&5
diff --git a/src/bignum.h b/src/bignum.h
index e59ab6b..5959ace 100644
--- a/src/bignum.h
+++ b/src/bignum.h
@@ -31,17 +31,17 @@ private:
   std::unique_ptr<char_t[]> buf_;
 
 public:
-  inline ulong() : buf_(make_unique<char_t[]>(dim)) {}
-  inline ulong(size_t t) : buf_(make_unique<char_t[]>(dim))
+  inline ulong() : buf_(std::make_unique<char_t[]>(dim)) {}
+  inline ulong(size_t t) : buf_(std::make_unique<char_t[]>(dim))
   {
     memcpy(buf_.get(), (char_t*)&t, sizeof(t));
   }
-  inline ulong(const ulong<dim>& rhs) : buf_(make_unique<char_t[]>(dim))
+  inline ulong(const ulong<dim>& rhs) : buf_(std::make_unique<char_t[]>(dim))
   {
     memcpy(buf_.get(), rhs.buf_.get(), dim);
   }
   explicit inline ulong(const char_t* data, size_t size)
-      : buf_(make_unique<char_t[]>(dim))
+      : buf_(std::make_unique<char_t[]>(dim))
   {
     if (size > dim) {
       throw std::bad_alloc();
```
But I assume a proper fix requires m4 changes of some kind.

**Comment by Polynomial-C** [comment]
This build failure also happens on Linux.

**Comment by tatsuhiro-t** [comment]
This bug should be fixed by e8e04d6f22a507e8374651d3d2343cd9fb986993.
The workaround is build aria2 with one of OpenSSL, gmp, or gcrypt.

**Comment by mebbert** [comment]
I also had trouble with this on Darwin and CentOS. I had downloaded the latest release (1.34.0), not realizing it was still the same release this bug existed in. Make sure to pull from the latest source (not the latest release). 

Also, on Darwin, you can get aria2 through homebrew. If you also want it to be built `--with-libssh2`, then you need to tell homebrew to compile from source.
